src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+
+pub const DEFAULT_BUF_SIZE: usize = 64 * 1024;
+
+
+pub fn read_to_end_zeroed<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> io::Result<usize> {
+    unsafe { read_to_end(r, buf, true) }
+}
+
+pub unsafe fn read_to_end_uninitialized<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>)
+                                                          -> io::Result<usize> {
+    read_to_end(r, buf, false)
+}
+
+// Used to fill a Vec from a Read type until there's no more data to come.
+// Uses an adaptive system to extend the vector when it fills. We want to
+// avoid paying to allocate and zero a huge chunk of memory if the reader only
+// has 4 bytes while still making large reads if the reader does have a ton
+// of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
+// time is 4,500 times (!) slower than this if the reader has a very small
+// amount of data to return.
+// function is unsafe because it can (optionally) work with uninitialised
+// buffers by passing zero_buf=false.  If using zero_buf=false, users should
+// ensure that their Read implementation does not read data from the buffer,
+// and always returns the correct number of bytes read.  Otherwise, behaviour
+// is undefined.
```

Comment:
Could you move this paragraph to the `read_to_end_uninitialized` function? Also, perhaps rewording it slightly to something like:

> This function is unsafe because it can optionally call the underlying `read` function with a slice into uninitialized memory. The default implementation of `read_to_end` for readers will zero out new memory in the `buf` before passing it to `read`, but avoiding this zero can often lead to a fairly significant perfomance win.
>
> Implementations using this method have to adhere to two guarantees:
>
> 1. The implementation of `read` never reads the buffer provided.
> 2. The implementation of `read` correctly reports how many bytes were written.